Change delete rule to soft delete

molly-moen opened the pull request:
# Type of change? <!-- [x] all the boxes that apply -->

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation <!-- Please help us reviewing your PR -->
Change delete rule API to mark rule as deleted instead of deleting the rule. Also change update rule method to fail if the rule has been deleted.

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

ppathan commented on Services/Rules.cs:
```diff
@@ -252,7 +260,11 @@ public class Rules : IRules
+            }
+            else if(savedRule.Deleted)
+            {
+                throw new InvalidInputException("Rule has been deleted. Cannot update.");
```
We should throw NotFound.

ppathan commented on Services/Rules.cs:
```diff
@@ -85,13 +85,21 @@ public class Rules : IRules
             {
-                this.log.Debug("id contains illegal characters.", () => new { id });
-                throw new InvalidInputException("id contains illegal characters.");
+                throw new InvalidInputException("Rule already deleted");
```
We should throw NotFound.

jillcary commented on Services/Rules.cs:
```diff
@@ -85,13 +85,21 @@ public class Rules : IRules
             {
-                this.log.Debug("id contains illegal characters.", () => new { id });
-                throw new InvalidInputException("id contains illegal characters.");
+                throw new InvalidInputException("Rule already deleted");
```
This could instead return a 204 No Content perhaps, since the operation the user has been requesting already happened. Either that or Not Found.

molly-moen replied:
I think not found will be clearest

jillcary commented on Services.Test/RulesTest.cs:
```diff
@@ -48,15 +54,81 @@ public RulesTest()
+        }
+
+        /**
+         * Verify if rule is already deleted and delete is called it will throw InvalidInputException
```
This comment reads strangely, perhaps rephrase to:
Verify if rule is already deleted and **when** delete is called it will throw InvalidInputException

jillcary commented on Services.Test/RulesTest.cs:
```diff
@@ -48,15 +54,81 @@ public RulesTest()
+                ETag = "123"
+            };
+            this.SetUpStorageAdapterGet(test);
+
```
nit: add comments for // Act //Assert

molly-moen replied:
done

molly-moen commented:
Asa manager doesn't need an update because it checks if rules are disabled, and deleting a rule also sets enabled to false. I will send java PR after this one

molly-moen commented:
Added ignore delete to asa manager, PR here: https://github.com/Azure/asa-manager-dotnet/pull/24

dluc commented on Services/Rules.cs:
```diff
@@ -85,13 +86,21 @@ public class Rules : IRules
             {
-                this.log.Debug("id contains illegal characters.", () => new { id });
-                throw new InvalidInputException("id contains illegal characters.");
+                throw new ResourceNotFoundException($"Rule {id} not found");
```
delete should fail silently when a resource doesn't exist or is already deleted - reporting OK

dluc replied:
ping

molly-moen replied:
already fixed, I think you are looking at outdated code

dluc commented on Services/Rules.cs:
```diff
@@ -85,13 +86,21 @@ public class Rules : IRules
 
-            await this.storage.DeleteAsync(STORAGE_COLLECTION, id);
+            existing.Deleted = true;
+            existing.Enabled = false;
```
I don't think this is required, and it would be nice to leave resources untouched to know what their status was before deletion

dluc replied:
ping: please remove

molly-moen replied:
removed, I think you are looking at outdated code

dluc commented on Services/Rules.cs:
```diff
@@ -115,7 +124,8 @@ public class Rules : IRules
+            string groupId,
+            bool includeDeleted)
         {
             var data = await this.storage.GetAllAsync(STORAGE_COLLECTION);
```
I don't remember the internals, but it's possible that this call returns only a subset of the all the alarms, expecting the client to paginate through the full list.  If that's the case the code could get a page a deleted alarms and eventually return an empty list when `includeDeleted` is `false`.  I wonder if the `includeDeleted` parameter should passed through to the storage, and query the original storage filtering out deleted records at the source

molly-moen replied:
we currently cannot pass the includeDeleted parameter to storage because the rule data is written as a string, not a json object. Adding this would require changing how we write rules to cosmos db, which would probably be a breaking change. But yes the scenario you describe is possible

dluc replied:
What's the solution for returning all the rules when GetAllAsync returns only 1 page of many?

molly-moen replied:
Actually on digging further this scenario is not possible. This query goes to the storage adapter, which queries cosmos db using CreateDocumentQuery....AsList(), which per their docs:
" 1.4 - Query for documents by a property other than Id
 NOTE: Operations like AsEnumberable(), ToList(), ToArray() will make as many trips to the database as required to fetch the entire result-set."
https://github.com/Azure/azure-documentdb-dotnet/blob/f374cc601f4cf08d11c88f0c3fa7dcefaf7ecfe8/samples/code-samples/DocumentManagement/Program.cs#L248-L251

dluc replied:
interesting, we should document this behavior in the storage adapter and test for it, in case we refactor that code. i.e. that was never a feature of the storage adapter, but now this code depends on that behavior

dluc commented on Services/Rules.cs:
```diff
@@ -252,7 +263,11 @@ public class Rules : IRules
+            }
+            else if(savedRule.Deleted)
+            {
+                throw new ResourceNotFoundException($"Rule {rule.Id} not found");
```
this doesn't seem right. If the rule is soft deleted, the creation should happen

dluc replied:
if Upsert is used only for Update then we should rename the method.  The word "Upsert" means that the object is Updated if it exists, or Inserted if it doesn't.

dluc commented on WebService/v1/Models/RuleApiModel.cs:
```diff
@@ -58,6 +60,9 @@ public class RuleApiModel
         };
 
+        [JsonProperty(PropertyName = "Deleted")]
+        public bool Deleted { get; set; }  = false;
```
soft deletion is an implementation detail. This property should be exposed only if the client is asking to include deleted items in the response. And I think it can be confusing seeing a list of "non deleted" rules when making a normal request

dluc commented:
> Asa manager doesn't need an update because it checks if rules are disabled, and deleting a rule also sets enabled to false. I will send java PR after this one

@molly-moen in ASA manager I wouldn't rely on this behavior, it couples the two implementations too much. ASA manager fetches the list of rules from this microservice, which should not return deleted rules.

molly-moen commented:
@dluc already added ignore delete to asa manager as well after talking with Parvez yesterday, PR: https://github.com/Azure/asa-manager-dotnet/pull/24

dluc commented:
@molly-moen so when deleting a rule it's not required to flip the state to disabled, correct?

molly-moen commented:
@dluc yes I removed the state flip in this iteration

dluc commented on WebService/v1/Controllers/RulesController.cs:
```diff
@@ -45,7 +46,8 @@ public RulesController(IRules ruleService)
                     groupId,
-                    includeDeleted.Value));
+                    includeDeleted.Value),
+                includeDeleted.Value);
```
indentation

molly-moen replied:
this is actually correct indentation:
 return new RuleListApiModel(
                await this.ruleService.GetListAsync( //1st parameter
                    order, //parameters for 1st parameter
                    skip.Value, //...
                    limit.Value, //...
                    groupId, //...
                    includeDeleted.Value), //...
                includeDeleted.Value); //2nd parameter

ppathan commented on Services/Rules.cs:
```diff
@@ -127,8 +136,9 @@ public class Rules : IRules
-                    if (string.IsNullOrEmpty(groupId) ||
+                    if ((string.IsNullOrEmpty(groupId) ||
                         rule.GroupId.Equals(groupId, StringComparison.OrdinalIgnoreCase))
+                        && (includeDeleted || !rule.Deleted))
```
nit: I would flip the order `&& (!rule.Deleted || includeDeleted))` Just to be clear that if rule is not deleted it will be included be default.

dluc commented on Services/Rules.cs:
```diff
@@ -242,17 +252,28 @@ public class Rules : IRules
 
+            if (savedRule != null && savedRule.Deleted)
+            {
+                throw new ResourceNotFoundException($"Rule {rule.Id} not found");
```
when this happens, line 260 should not be executed, otherwise the log will be wrong

molly-moen replied:
line 260 won't get executed, that would mean savedRule == null (if the rule is not found it can't be "deleted")

dluc replied:
gotcha, thanks

molly-moen commented:
@Devis currently it throws an error, I will fix it to return 200

dluc commented on Services/Rules.cs:
```diff
@@ -86,7 +86,17 @@ public class Rules : IRules
+            {
+                existing = await this.GetAsync(id);
+            }
+            catch (Exception exception)
```
1. please add tests
2. only ResourceNotFoundException should have this behavior. E.g. if delete fails because the storage is not available, the API should return 500

molly-moen commented:
Fixed status codes. There are currently no tests for any controllers in telemetry service (the only one that exists is commented out), so adding controller tests is out of scope, especially since this pr has been open for 2 weeks. I have added lots of tests on rules class.
